### ISSUE TITLE
[API] ConvolutionalEncoder : Some confusion about the ConvolutionalEncoder interface.

### ISSUE THREAD

**Issue by vanewu** [dmlc/gluon-nlp#471]
This interface is very good, but there are two points that are confusing.
 First, this interface does not provide padding. Although the effect of padding may be small in one-dimensional convolution due to GlobalMaxPool1D. But acutally it may be a little different. If I have a specific demand that want to pad data, I need to pad the input 3D data independently. But the ndarray padding currently only supports 4d and 5d. I need to change the dimension, which adds some extra unnecessary operations. Maybe adding padding to the implementation of this interface will be easier to use. Also adding padding to this interface is easy because conv1d is used internally.
```python
seq.add(nn.Conv1D(in_channels=self._embed_size,
                                      channels=num_filter,
                                      kernel_size=ngram_size,
                                      use_bias=True))
```
Second, I think the input of the general convolution will be the 0th dimension of the batch, but the input of this interface is not. I have seen the implementation. it has internally performed the dimensional transformation. Is this necessary?
```python
inputs = F.transpose(inputs, axes=(1, 2, 0))
``` 
Maybe this interface will be more intuitive when the batch is in the 0th dimension.

**Comment by cgraywang** [comment]
@kenjewu thanks for your interests. 

For question 1, the initial design of this API is based on the simple, clear and DATA ALREADY BEEN PADDED assumption, which is similar to many other model API in GluonNLP. It is actually simple to use API in the batchify API including the [PAD operator](https://github.com/dmlc/gluon-nlp/blob/bcae0fef59d31d9524c9bc513852b2e897a5e0b2/src/gluonnlp/data/batchify/batchify.py#L160) in GluonNLP. Could you please try the PAD the data first and then handle the PAD in the loss if necessary?

For question 2, since the `ConvolutionEncoder` API is mostly used in the embedding layer of many seq2seq model, especially providing the character embedding. The seq2seq model in the NLP mostly requires the input in the layout of `TNC`,  besides as you know, the `Conv` operators requires the first dimension to be N (batch_size) and second dimension to be C (channel), so it is necessary to have this in the API.

Please feel free to reach out if you have more questions.

**Comment by kenjewu** [comment]
@cgraywang Thank you very much for your reply. 

According to your explanation, I have no doubt about the second question.  Thank you!

But for the first problem, if I pre-populate the data, it seems that I can't consider all the situations, because if I use multiple different convolution kernels, the size of the pad for each kernel is different.  At this point, using the PAD operation in the batchify seems a little cumbersome, and it seems more straightforward to provide padding parameters in Conv1D for each kernel at this location. Maybe padding doesn't have much impact, I'm not sure about it. But I recommend adding the padding parameter to improve generality.
